### PR TITLE
FE-1 Did you award a contract for 'Buying some hosting'? page details

### PR DESCRIPTION
**Pull request by lfdebrux**
[Trello](https://trello.com/c/yIUrcLLo/73-fe-1-did-you-award-a-contract-for-buying-some-hosting-page-details)

1. Added "Did you award a contract?" page
2. Created a `DidYouAwardAContractForm` class for validating the radio input
3. Created a test suite `TestDirectAwardAwardContract` for the page, HTML form, and redirects
4. Created a route function `award_contract` that passes the tests

Currently as there are no routes for `which-service-won-contract` and `why-didnt-you-award-contract` the tests for those have been passed by sliming.

Dealing with the "We are still assessing services" flow is in ticket [#110](https://trello.com/c/5Lrp1lOK/110-fe-8-we-are-still-assessing-flow)

### CODE REVIEW

**Comment by lfdebrux** [comment]
Thanks @DilwoarH, I've addressed all your changes except the content change

File: app/main/views/g_cloud.py [hunk 1]
```diff
@@ -544,6 +544,62 @@ def end_search(framework_framework, project_id):
...
+            return redirect(url_for('.which_service_won_contract',
+                                    framework_framework=framework_framework,
+                                    project_id=project_id))
+        elif form.did_you_award_a_contract.data == 'no':
+            return redirect(url_for('.why_didnt_you_award_contract',
+                                    framework_framework=framework_framework,
+                                    project_id=project_id))
+        else:
+            # this else should never be reached
+            abort(500, "Unexpected answer to 'Did you award contract?'")
```
Comment (galund) [inline app/main/views/g_cloud.py]: I know Dilwoar mentioned this already, I think if you want to leave it in I'd be find with not bothering to have a message, `abort(500)` would be fine.

File: app/main/views/g_cloud.py [hunk 2]
```diff
@@ -544,6 +544,62 @@ def end_search(framework_framework, project_id):
...
+@direct_award.route(
+    '/<string:framework_framework>/projects/<int:project_id>/which-service-won-contract',
+    methods=['GET']
+)
+def which_service_won_contract(framework_framework, project_id):
+    pass
+
+
+@direct_award.route(
+    '/<string:framework_framework>/projects/<int:project_id>/why-didnt-you-award-contract',
```
Comment (galund) [inline app/main/views/g_cloud.py]: So I know it's kinda late to be mentioning this, but I really don't like URI's that look like they are missing punctuation. `.../non-award-reason` perhaps? cc @karlchillmaid

Comment (lfdebrux) [reply app/main/views/g_cloud.py]: Response from @karlchillmaid on Slack

> @george.lund @dilwoar.hussain re the URL question on GitHub — the URLs follow the standards for GOV.UK: https://insidegovuk.blog.gov.uk/url-standards-for-gov-uk/
The guidelines don't say anything about contractions, but when we discussed this at content crit we found other examples that do, eg. https://www.gov.uk/expenses-if-youre-self-employed
I agree that contractions look a bit funny without the apostrophe, but I think they're still readable, and it's probably more important to keep the URLs as close as possible to the page title and use logical, consistent, predictable patterns for URLs.

File: app/main/forms/direct_award_forms.py [hunk 3]
```diff
@@ -14,3 +15,14 @@ class CreateProjectForm(FlaskForm):
             Optional()
         ]
     )
+
+class DidYouAwardAContractForm(FlaskForm):
+    did_you_award_a_contract = RadioField(
+        "Did you award a contract?",
+        validators=[InputRequired(message="You need to answer this question.")],
```
[the remark this answers is not in the document]
Comment (lfdebrux) [reply app/main/forms/direct_award_forms.py]: I talked to @karlchillmaid about this this morning, he said that since the phrase is copied from another page he was okay with us keeping it as is for now for consistency.

However, he also said that it was a good question, and was going to bring it up at content crit to see if people had any thoughts on error messages, and possibly we will revisit this question across Digital Marketplace at a later time.

tl;dr let's merge it!

File: app/templates/direct-award/did-you-award-contract.html [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+<div class="did-you-award-contract-page">
+  <div class="grid-row">
+    <div class="column-two-thirds">
+      <header class="page-heading">
+        <h1>
+          Did you award a contract for '{{project.name}}'?
+        </h1>
+      </header>
+
+      <form method="POST" action="{{ url_for('direct_award.did_you_award_contract', framework_framework=framework.framework, project_id=project.id) }}">
```
Comment (galund) [inline app/templates/direct-award/did-you-award-contract.html]: It may be possible to use `action=""` and if so that would probably be preferable?

Comment (lfdebrux) [reply app/templates/direct-award/did-you-award-contract.html]: According to [SO](https://stackoverflow.com/questions/1131781/is-it-a-good-practice-to-use-an-empty-url-for-a-html-forms-action-attribute-a) `action=""` _can_ cause issues.

I agree it is nicer though not to have a url in there. Is there any reason not just to leave the `action` attribute out altogether? Thoughts @DilwoarH?

Comment (DilwoarH) [reply app/templates/direct-award/did-you-award-contract.html]: https://stackoverflow.com/a/9401608

Have a read of the above, I personally would leave it as it is as there may be some browsers which behave differently without an action. It's also standard to have an action target - removing it won't bring any additional benefit.

Comment (lfdebrux) [reply app/templates/direct-award/did-you-award-contract.html]: Okay, I've left it as it is, but also added a test `test_award_contract_form_action_url_is_award_contract_url` that checks that the `action` url is the same as the GET url

**Comment by benvand** [comment]
Routes that don't return a HTTP response cause 500s in flask. We should back this out. It's blocking the pipeline because it can't go to production as is.
https://github.com/alphagov/digitalmarketplace-buyer-frontend/pull/760

File: app/main/views/g_cloud.py [hunk 5]
```diff
@@ -544,6 +544,61 @@ def end_search(framework_framework, project_id):
...
+)
+def did_you_award_contract(framework_framework, project_id):
+    all_frameworks = data_api_client.find_frameworks().get('frameworks')
+    framework = framework_helpers.get_latest_live_framework(all_frameworks, framework_framework)
+
+    # Get the requested Direct Award Project.
+    project = data_api_client.get_direct_award_project(project_id=project_id)['project']
+
+    form = DidYouAwardAContractForm()
+    form.did_you_award_a_contract.options = [{"label": label, "value": value}
```
Comment (benvand) [inline app/main/views/g_cloud.py]: Dynamically editing the form object like this should be done in the `__init__` method of the form. The `.options` (or historically `.toolkit_macro_options`) attribute is something we're going to need every time we instantiate this form
